Fix roots_quintic

ForeverHaibara opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
Add support in `polys.polyroots.roots_quintic` to solve quintics with nonzero degree-4 coefficient by applying a linear transform. In the past, it only returned an empty list in this case.

Also fix a small bug in this function that it failed to solve quintics like 'x^5-2'. This was caused by miswriting `if r2 is not None` as `if r2`.

#### Other comments
Another tricky quintic that it fails to solve is presented below, which I guess is due to numerical instability. Might get it fixed in the future.
```python3
sp.polys.polyroots.roots_quintic(sp.sympify('x^5-110*x^3-55*x^2+2310*x+979').subs('x','3*x/2').as_poly()) # fails
sp.polys.polyroots.roots_quintic(sp.sympify('x^5-110*x^3-55*x^2+2310*x+979').subs('x','5*x/4').as_poly()) # good
```

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

* polys
  * Made `roots_quintic()` accept quintics in general form (nonzero degree-4 coeff is now okay)

<!-- END RELEASE NOTES -->

ForeverHaibara committed:
Fix roots_quintic

Add support for quintic roots solver to solve quintic with nonzero degree 4 coefficient. Also fix a small bug that it fails to solve quintics like 'x^5-2'.

smichr commented on sympy/polys/polyroots.py:
```diff
@@ -504,33 +504,41 @@ def roots_cyclotomic(f, factor=False):
     if not f.is_irreducible:
         return result
-
+        
```
```suggestion

```

smichr commented on sympy/polys/polyroots.py:
```diff
@@ -504,33 +504,41 @@ def roots_cyclotomic(f, factor=False):
     if f20.is_irreducible:
         return result
-
+        
```
```suggestion

```

smichr committed:
pep8

skieffer commented:
@ForeverHaibara thanks for this PR! Haven't reviewed closely yet, but one important thing will be to add at least one new unit test, showing something that works now but would not have worked before. `sympy/polys/tests/test_polyroots.py` would be a good place for that.

smichr commented on sympy/polys/polyroots.py:
```diff
@@ -504,33 +504,39 @@ def roots_cyclotomic(f, factor=False):
 def roots_quintic(f):
     """
-    Calculate exact roots of a solvable quintic
+    Calculate exact roots of a solvable irreducible quintic with rational coefficients. 
```
```suggestion
    Calculate exact roots of a solvable irreducible quintic with rational coefficients.
```

smichr committed:
pep8-2

ForeverHaibara committed:
add roots_quintic tests

add some irreducible quintics to test roots_quintic

ForeverHaibara commented:
> @ForeverHaibara thanks for this PR! Haven't reviewed closely yet, but one important thing will be to add at least one new unit test, showing something that works now but would not have worked before. `sympy/polys/tests/test_polyroots.py` would be a good place for that.

Thanks! I have now added `test_roots_quintic` in the file `sympy/polys/tests/test_polyroots.py`.

ForeverHaibara committed:
pep8

fix trailing whitespace

ForeverHaibara committed:
check if univariate in roots_quintic

add check in roots_quintic to test the quintic is univariate, return empty list otherwise

ForeverHaibara committed:
fix irrational coefficient cases in root_quintics

fix irrational coefficient cases in root_quintics

skieffer commented on sympy/polys/polyroots.py:
```diff
@@ -504,33 +504,42 @@ def roots_cyclotomic(f, factor=False):
-
-    # Eqn must be of the form x^5 + px^3 + qx^2 + rx + s
-    if coeff_4:
+    if not f.is_univariate:
         return result
```
Whether to check things like this raises the usual questions around public vs private functions, and GIGO.

The name of `roots_quintic()` does not begin with a `_`, which tends to make it look like a public function, but I don't find this function in the API reference either, which tends to make it look private. I don't know if we've settled the policy on this yet -- see #23037.

I suspect that up to now `roots_quintic()` has largely been used only by the (explicitly public) `roots()` function, and that this is why the `r2 is not None` check below was not caught before now.

Is there a reason to call `roots_quintic()` directly, instead of just calling `roots()`?

If not, then maybe checks like these aren't necessary. If so, then the function should have a better docstring, and maybe even more such checks, like checking that `f` is actually degree 5, so that the `coeff_5, coeff_4, p_, q_, r_, s_ = f.all_coeffs()` line can't fail with a `ValueError`.

ForeverHaibara replied:
You are right. I think I shall remove this line. Other functions like `roots_cubic` does not check this either.

skieffer commented on sympy/polys/polyroots.py:
```diff
@@ -504,33 +504,42 @@ def roots_cyclotomic(f, factor=False):
-            return result
-        f = Poly(f/coeff_5)
-    elif not all(coeff.is_Rational for coeff in (p, q, r, s)):
+        f = Poly(f / coeff_5, extension = True)
+        _, coeff_4, p_, q_, r_, s_ = f.all_coeffs()
```
I understand setting `extension=True` is meant to handle cases where the coeffs are irrational, but here again I feel we need to clarify the API. The docstring currently says `f` is expected to have rational coefficients. Does this method in fact work when the coeffs are irrational? (Asking because I don't know.) As used by the `roots()` function, do we know `f` will always have rational coeffs by the time it gets here?

ForeverHaibara replied:
I suppose that it could also handle cases where the quintic is a multiple of a rational one, say `sqrt(2)*(x**5-2)`, though not explicitly informed in the docstring. But maybe it seems a little redundant. If this should be removed, then one should check whether the coeffs are rational here before calling `f.all_coeffs()` since there is a test passing in irrational quintics.

skieffer replied:
Ah, I see. You're referring to `solvers/tests/test_solvers.py:test_quintics_3`, which shows we actually do expect to sometimes receive polys `f` with irrational coeffs here. This seems reasonable then. But I would add a comment for the sake of future devs, explaining that this allows polys of the form you mention (i.e. constant multiples of rational ones) to pass.

skieffer commented:
The failing "authors" test just means you need to add yourself to the `.mailmap` file. See
<https://github.com/sympy/sympy/wiki/Development-workflow#add-your-name-and-email-address-to-the-mailmap-file>

skieffer commented:
Math and unit tests look good. Code quality really nice. So, apart from my questions around the API, this is basically looking good to me.

ForeverHaibara committed:
Update .mailmap

ForeverHaibara committed:
remove univariate check in roots_quintic

skieffer commented:
It seems to be hanging on this doctest:

https://github.com/sympy/sympy/blob/a0fa401b0a3db402a4ad740f14674560aba78f19/sympy/solvers/solvers.py#L726-L730

ForeverHaibara commented:
> It seems to be hanging on this doctest:
> 
> https://github.com/sympy/sympy/blob/a0fa401b0a3db402a4ad740f14674560aba78f19/sympy/solvers/solvers.py#L726-L730

Seems so. It passes into the roots_quintic a polynomial 

`Poly((-1)**(2/3)*_x0**5 - (-1)**(2/5)*_x0**3 + 3**(1/3) + (-1)**(3/5)*3**(1/5), _x0, domain='ZZ((-1)**(1/3),(-1)**(1/5),3**(1/3),3**(1/5))')`

 and `f = Poly(f / coeff_5, extension = True)` gets stuck there. I am not sure why it takes such long time to turn it into a polynomial with `extension = True`. Perhaps the simplest way to get around is to drop the support of, as mentioned earlier, constant multiples of rational polys.

oscarbenjamin commented:
It's just because `primitive_element` is slow at creating an extension field in this case. That in turn is because gcd over extension fields is slow.

ForeverHaibara committed:
roots_quintic check rational coeff at start

skieffer commented:
LGTM. Will merge if no more comments.

skieffer commented:
Thanks for the contribution, @ForeverHaibara !
